homeassistant/components/history/__init__.py

```diff
@@ -107,8 +107,7 @@ async def get(
         no_attributes = "no_attributes" in request.query
 
         if (
-            (end_time and not has_recorder_run_after(hass, end_time))
```

Comment:
I was hoping has_recorder_run_after would get fixed instead of dropping the check.

Reply:
The root cause of this issue is probably not in this code. There is something happening in recorder that loses previous runs in the database. What could be causing this is beyond my comprehension.

Reply:
Is the recorder run missing from the table? Maybe the purge code deletes it too early, and that’s what we need to fix instead?

Reply:
Absolutely, but really no idea what causes this to happen. It happens less than once per month and makes it hard to debug for me

Reply:
Here is the logic to remove the recorder runs in case you see anything obviously wrong
https://github.com/home-assistant/core/blob/55eb11055ca8a5fbb2109fda1f9319e7d8adbe99/homeassistant/components/recorder/purge.py#L581

Reply:
I think it shouldn't check the  start-time of the recorder_run, but the end_time.
https://github.com/home-assistant/core/blob/95eefbac20f683016367b76faed420369d675e58/homeassistant/components/recorder/queries.py#L611

Details here:
https://github.com/home-assistant/core/issues/117793#issuecomment-2456540900

Reply:
I'll take a look when I get back to the US early next week

Reply:
New PR: <https://github.com/home-assistant/core/pull/130378>

Reply:
The PR won't fix databases which have recorder runs already purged.

However, I'm not sure we really need to check the recorder runs table, why don't we instead just compare the end time in the requested history period with the purge settings?

Reply:
From Discord chat:
We call `has_recorder_run_after` in the history API as a way to avoid DB accesses if the recorder can't have such old events. However, checking the recorder runs is not a great way to determine that unless you restart Home Assistant frequently.
Instead, we should implement a new function named `has_events_before(timestamp: datetime) -> bool` or something, which returns `False` if `timestamp` is old enough that we should have purged the events already (based on purge settings)

Reply:
Maybe because you don't want to assume HA was running at the time?

Reply:
Right, if we consider only the timestamp + purge settings, we will not search for events which should be purged according to purge settings even if we've not purged them. In most cases, it will be more accurate than looking at the oldest recorder run though.
If we *really* want to be correct, we could store the newest purge date in the database instead, but it seems a bit unnecessary.

WDYT @bdraco ?

Reply:
For most its fine to calculate when its expected to be purged based on timestamp + purge settings and use that.

The corner cases to consider:
- auto purge is disabled
- purge is done via an automation instead of auto purge

So we probably need to store the purge date in the databases as the amount of people that use custom purge settings is probably enough to consider accomodating.

Reply:
OK, makes sense.

I think a purge runs should have the following columns:
- id
- purge run timestamp (is it enough to record when the purge finished, or do we need also the start time of the purge run?)
- purge_before timestamp

Purging the purge run table should be sure to preserve the row with the youngest `purge_before` timestamp

Reply:
When the purge is finished we need to store:

The timestamp of the possible oldest row in the database (ie purge before) in the table
Update the memory cache of the same

Reply:
OK, so we add a new row only when a purge completes?

Reply:
Yes

Reply:
PR https://github.com/home-assistant/core/pull/130378 has been merged, which corrects the logic for purging recorder runs. I've started working on the purge_runs table, but I'm not sure it'll be ready in time for HA Core 2024.12.

Reply:
After some more discussion, we don't think a  purge runs table is needed. Here's a PR which replaces checking recorder run start time with the timestamp of the oldest state in the database: https://github.com/home-assistant/core/pull/131702